File: Python/11-Math&Geometry/Problems/4-RotateImage/RotateImage.py
def rotate_image_approach1(matrix):
    """
    Approach 1: Using extra space for a new matrix.
    Creates a new matrix and copies elements from the original matrix in rotated order.

    Args:
        matrix: A list of lists representing the image.

    Returns:
        The rotated image (in-place modification of the input).
    """
    n = len(matrix)
    rotated = [[0] * n for _ in range(n)]  # Create a new matrix

    for i in range(n):
        for j in range(n):
            rotated[j][n - 1 - i] = matrix[i][j]

    # Copy the rotated matrix back to the original matrix
    for i in range(n):
        for j in range(n):
            matrix[i][j] = rotated[i][j]

    return matrix

def rotate_image_approach2(matrix):
    """
    Approach 2: In-place rotation by transposing and then reversing each row.
    Transpose the matrix (swap rows and columns) and then reverse each row.

    Args:
        matrix: A list of lists representing the image.

    Returns:
        The rotated image (in-place modification of the input).
    """
    n = len(matrix)

    # Transpose the matrix
    for i in range(n):
        for j in range(i + 1, n):
            matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

    # Reverse each row
    for row in matrix:
        row.reverse()

    return matrix

def rotate_image_approach3(matrix):
    """
    Approach 3: In-place rotation using four-way swaps.
    Iterate through layers and swap elements in four-way cycles.

    Args:
        matrix: A list of lists representing the image.

    Returns:
        The rotated image (in-place modification of the input).
    """
    n = len(matrix)

    for i in range(n // 2):
        for j in range(i, n - i - 1):
            temp = matrix[i][j]
            matrix[i][j] = matrix[n - 1 - j][i]
            matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j]
            matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i]
            matrix[j][n - 1 - i] = temp

    return matrix

def rotate_image_approach4(matrix):
    """
    Approach 4: Similar to approach 3, but more explicit variable names.
    Iterate through layers and swap elements in four-way cycles, with clearer variable names.

    Args:
        matrix: A list of lists representing the image.

    Returns:
        The rotated image (in-place modification of the input).
    """
    n = len(matrix)
    for layer in range(n // 2):
        first = layer
        last = n - 1 - layer
        for i in range(first, last):
            offset = i - first
            top = matrix[first][i]  # save top
            # left -> top
            matrix[first][i] = matrix[last - offset][first]
            # bottom -> left
            matrix[last - offset][first] = matrix[last][last - offset]
            # right -> bottom
            matrix[last][last - offset] = matrix[i][last]
            # top -> right
            matrix[i][last] = top
    return matrix

def rotate_image_approach5(matrix):
    """
    Approach 5: Using zip and reversed.
    Uses zip and reversed to create the rotated matrix.

    Args:
        matrix: A list of lists representing the image.

    Returns:
        The rotated image (in-place modification of the input).
    """
    n = len(matrix)
    rotated = list(zip(*matrix[::-1]))
    for i in range(n):
        for j in range(n):
            matrix[i][j] = rotated[i][j]
    return matrix

# Example usage and output
matrix1 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
matrix2 = [[5, 1, 9, 11], [2, 4, 8, 10], [13, 3, 6, 7], [15, 14, 12, 16]]

print("Original Matrix 1:", matrix1)
print("Approach 1 (Extra Space):", rotate_image_approach1(matrix1.copy()))
print("Approach 2 (Transpose + Reverse):", rotate_image_approach2(matrix1.copy()))
print("Approach 3 (Four-way Swaps):", rotate_image_approach3(matrix1.copy()))
print("Approach 4 (Four-way Swaps, clearer vars):", rotate_image_approach4(matrix1.copy()))
print("Approach 5 (zip and reversed):", rotate_image_approach5(matrix1.copy()))

print("\nOriginal Matrix 2:", matrix2)
print("Approach 1 (Extra Space):", rotate_image_approach1(matrix2.copy()))
print("Approach 2 (Transpose + Reverse):", rotate_image_approach2(matrix2.copy()))
print("Approach 3 (Four-way Swaps):", rotate_image_approach3(matrix2.copy()))
print("Approach 4 (Four-way Swaps, clearer vars):", rotate_image_approach4(matrix2.copy()))
print("Approach 5 (zip and reversed):", rotate_image_approach5(matrix2.copy()))

File: C++/11-Math&Geometry/Problems/4-RotateImage/RotateImage.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric> // for iota

using namespace std;

// Function to print the matrix
void printMatrix(const vector<vector<int>>& matrix) {
    for (const auto& row : matrix) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
    cout << endl;
}

// Approach 1: Using an auxiliary matrix (simplest, not in-place)
// - Create a new matrix with dimensions swapped (n x n becomes n x n).
// - Copy elements from the original matrix to the new matrix in rotated order.
// - Time Complexity: O(n^2), Space Complexity: O(n^2)
vector<vector<int>> rotateImage_approach1(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    vector<vector<int>> rotated(n, vector<int>(n)); // Initialize new matrix

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            rotated[j][n - 1 - i] = matrix[i][j];
        }
    }
    return rotated;
}

// Approach 2: In-place rotation with layers (most efficient)
// - Rotate the matrix layer by layer.
// - For each layer, rotate the four edges using a temporary variable.
// - Time Complexity: O(n^2), Space Complexity: O(1)
void rotateImage_approach2(vector<vector<int>>& matrix) {
    int n = matrix.size();
    for (int layer = 0; layer < n / 2; ++layer) {
        int first = layer;
        int last = n - 1 - layer;
        for (int i = first; i < last; ++i) {
            int offset = i - first;
            int top = matrix[first][i]; // save top

            // left -> top
            matrix[first][i] = matrix[last - offset][first];

            // bottom -> left
            matrix[last - offset][first] = matrix[last][last - offset];

            // right -> bottom
            matrix[last][last - offset] = matrix[i][last];

            // top -> right
            matrix[i][last] = top;
        }
    }
}

// Approach 3: Using reverse and transpose (in-place)
// - Transpose the matrix (swap rows and columns).
// - Reverse each row of the transposed matrix.
// - Time Complexity: O(n^2), Space Complexity: O(1)
void rotateImage_approach3(vector<vector<int>>& matrix) {
    int n = matrix.size();
    // Transpose the matrix
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            swap(matrix[i][j], matrix[j][i]);
        }
    }
    // Reverse each row
    for (int i = 0; i < n; ++i) {
        reverse(matrix[i].begin(), matrix[i].end());
    }
}

// Approach 4: Using nested loops and swapping (in-place, less optimized than 2)
// - Iterate through half of the matrix.
// - Swap elements in a circular manner for each element in the first half.
// - Time Complexity: O(n^2), Space Complexity: O(1)
void rotateImage_approach4(vector<vector<int>>& matrix) {
    int n = matrix.size();
    for (int i = 0; i < n / 2; i++) {
        for (int j = i; j < n - i - 1; j++) {
            int temp = matrix[i][j];
            matrix[i][j] = matrix[n - 1 - j][i];
            matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j];
            matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i];
            matrix[j][n - 1 - i] = temp;
        }
    }
}

// Approach 5: Using std::rotate (In-place, more concise, same efficiency as Approach 2)
// - Similar to approach 2, but uses the standard library function `std::rotate`
// - Time Complexity: O(n^2), Space Complexity: O(1)
void rotateImage_approach5(vector<vector<int>>& matrix) {
    int n = matrix.size();
    for (int layer = 0; layer < n / 2; ++layer) {
        int first = layer;
        int last = n - 1 - layer;
        for (int i = first; i < last; ++i) {
            int offset = i - first;
            vector<int*> ptrs = {
                &matrix[first][i],
                &matrix[last - offset][first],
                &matrix[last][last - offset],
                &matrix[i][last]
            };
            int temp = *ptrs[0];
            rotate(ptrs.rbegin(), ptrs.rbegin() + 1, ptrs.rend()); // Rotate pointers
            *ptrs[3] = temp;
        }
    }
}

int main() {
    // Example usage
    vector<vector<int>> matrix = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    cout << "Original Matrix:" << endl;
    printMatrix(matrix);

    // Approach 1
    vector<vector<int>> rotated1 = rotateImage_approach1(matrix);
    cout << "Rotated Matrix (Approach 1):" << endl;
    printMatrix(rotated1);

    // Approach 2 (in-place)
    vector<vector<int>> matrix2 = matrix; // Make a copy to preserve original
    rotateImage_approach2(matrix2);
    cout << "Rotated Matrix (Approach 2):" << endl;
    printMatrix(matrix2);

    // Approach 3 (in-place)
    vector<vector<int>> matrix3 = matrix;
    rotateImage_approach3(matrix3);
    cout << "Rotated Matrix (Approach 3):" << endl;
    printMatrix(matrix3);

    // Approach 4 (in-place)
    vector<vector<int>> matrix4 = matrix;
    rotateImage_approach4(matrix4);
    cout << "Rotated Matrix (Approach 4):" << endl;
    printMatrix(matrix4);

    // Approach 5 (in-place)
    vector<vector<int>> matrix5 = matrix;
    rotateImage_approach5(matrix5);
    cout << "Rotated Matrix (Approach 5):" << endl;
    printMatrix(matrix5);

    return 0;
}
